### PR TITLE
Core/Scripts: Move some spells to scripts

### PR DESCRIPTION
**Pull request by DanVS**
**Changes proposed:**
Move generic school damage spells (from `@todo`) to scripts. Also applied to four horsemen script.

**Target branch(es):** 3.3.5

**Tests performed:** Does build and partial tests shown that works in-game.

### CODE REVIEW

File: sql/updates/world/3.3.5/9999_99_99_99_world.sql [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+('spell_four_horsemen_consumption',
+'spell_rajaxx_thundercrash',
+'spell_gen_arcane_charge',
+'spell_pet_dk_gargoyle_strike');
+
+INSERT INTO `spell_script_names` VALUES
+(28865, 'spell_four_horsemen_consumption'),
+(25599, 'spell_rajaxx_thundercrash'),
+(45072, 'spell_gen_arcane_charge'),
+(51963, 'spell_pet_dk_gargoyle_strike');
```
Comment (Aokromes) [inline sql/updates/world/3.3.5/9999_99_99_99_world.sql]: missing newline :)

Comment (DanVS) [reply sql/updates/world/3.3.5/9999_99_99_99_world.sql]: Fixed. Thanks!

File: sql/updates/world/3.3.5/9999_99_99_99_world.sql [hunk 2]
```diff
@@ -0,0 +1,12 @@
+DELETE FROM `spell_script_names` WHERE `ScriptName` IN
+('spell_four_horsemen_consumption',
+'spell_rajaxx_thundercrash',
+'spell_gen_arcane_charge',
+'spell_pet_dk_gargoyle_strike');
+
+INSERT INTO `spell_script_names` VALUES
```
Comment (illfated) [inline sql/updates/world/3.3.5/9999_99_99_99_world.sql]: Codestyle: TC prefers that the column names are included in SQL queries:
```sql
INSERT INTO `spell_script_names` (`spell_id`,`ScriptName`) VALUES
```

Comment (DanVS) [reply sql/updates/world/3.3.5/9999_99_99_99_world.sql]: Fixed. Thanks!

**Comment by illfated** [comment]
Other than that, it looks like you need to rebase your branch `scripts` on TC commit https://github.com/TrinityCore/TrinityCore/commit/829527a08ef2432c6c0ff5e65d8012dc396d155a (current) to get rid of the fails in AppVeyor and travis-ci.

**Comment by DanVS** [comment]
Rebased. Seems that checks do need manual restart.

**Review by ghost — APPROVED** [review]
travis-ci & AppVeyor problems solved.
Real coders needed to review the content.

File: src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp [hunk 3]
```diff
@@ -690,70 +690,77 @@ class boss_four_horsemen_sir : public CreatureScript
...
-        class spell_four_horsemen_mark_AuraScript : public AuraScript
-        {
-            PrepareAuraScript(spell_four_horsemen_mark_AuraScript);
+ class spell_four_horsemen_mark : public AuraScript
+ {
+     PrepareAuraScript(spell_four_horsemen_mark);
+
+     void OnApply(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
+     {
+         if (Unit *caster = GetCaster())
```
Comment (Jildor) [inline src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp]: ```Unit* caster```

File: src/server/scripts/Pet/pet_dk.cpp [hunk 4]
```diff
@@ -136,8 +136,31 @@ class npc_pet_dk_guardian : public CreatureScript
...
+class spell_pet_dk_gargoyle_strike : public SpellScript
+{
+    PrepareSpellScript(spell_pet_dk_gargoyle_strike);
+
+    void HandleDamageCalc(SpellEffIndex /*effIndex*/)
+    {
+        int32 damage = 60;
+        if (Unit* caster = GetCaster())
+        {
+            if (GetCaster()->getLevel() >= 60)
```
Comment (Jildor) [inline src/server/scripts/Pet/pet_dk.cpp]: Replace ```GetCaster()``` by ```caster```

File: src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp [hunk 5]
```diff
@@ -690,70 +690,77 @@ class boss_four_horsemen_sir : public CreatureScript
...
 
-            void Register() override
-            {
-                AfterEffectApply += AuraEffectApplyFn(spell_four_horsemen_mark_AuraScript::OnApply, EFFECT_0, SPELL_AURA_DUMMY, AURA_EFFECT_HANDLE_REAL_OR_REAPPLY_MASK);
-            }
-        };
+     void HandleDamageCalc(SpellEffIndex /*effIndex*/)
+     {
+         SetHitDamage(2750);
+         if (GetCaster()->GetMap()->IsHeroic())
```
Comment (Jildor) [inline src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp]: can crash here?

Comment (DanVS) [reply src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp]: Doubt this.

File: src/server/scripts/Pet/pet_dk.cpp (hunk 4 above)
Comment (DanVS) [reply src/server/scripts/Pet/pet_dk.cpp]: Fixed. Thanks!

File: src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp (hunk 3 above)
Comment (DanVS) [reply src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp]: Fixed. Thanks!

File: src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp (hunk 5 above)
Comment (Shauren) [reply src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp]: nope. caster is never null in SpellScript and nothing can cast a spell without being on map

File: src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp [hunk 6]
```diff
@@ -690,70 +690,77 @@ class boss_four_horsemen_sir : public CreatureScript
...
+     PrepareSpellScript(spell_four_horsemen_consumption);
 
-            void Register() override
-            {
-                AfterEffectApply += AuraEffectApplyFn(spell_four_horsemen_mark_AuraScript::OnApply, EFFECT_0, SPELL_AURA_DUMMY, AURA_EFFECT_HANDLE_REAL_OR_REAPPLY_MASK);
-            }
-        };
+     void HandleDamageCalc(SpellEffIndex /*effIndex*/)
+     {
+         SetHitDamage(2750);
```
Comment (Keader) [inline src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp]: not best check if is heroic first?
like:

```cpp
if (GetCaster()->GetMap()->IsHeroic())
    SetHitDamage(4250);
else
    SetHitDamage(2750);
```
instead call SetHitDamage twice?

Comment (Jildor) [reply src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp]: Could it be better like this?
```c++
SetHitDamage((GetCaster()->GetMap()->IsHeroic()) ? 4250 : 2750);
```

Comment (Keader) [reply src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp]: is fine to me too.
But still is better put it in a variable first.
like:

```cpp
uint32 damage = GetCaster()->GetMap()->IsHeroic() ? 4250 : 2750;
SetHitDamage(damage);
```

Comment (Jildor) [reply src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp]: But in this case with uint16 it's enough 😄

Comment (Keader) [reply src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp]: yeah, but we always use int32/uint32 to represent damages xD

Comment (DanVS) [reply src/server/scripts/Northrend/Naxxramas/boss_four_horsemen.cpp]: Fixed. Thanks!

**Comment by ccrs** [comment]
missing spellscript header in a couple of them

**Comment by DanVS** [comment]
Headers appended. Thanks!

**Comment by jackpoz** [comment]
this is causing https://github.com/TrinityCore/TrinityCore/issues/21435

**Pull request merged**